File: server/internal/transactions/db/tags.go
package transactionsdb

import (
	"context"
	"regexp"

	"tallyo/internal/apierror"
	"tallyo/internal/database/dbgen"
	"tallyo/internal/database/dbutil"
	"tallyo/internal/graph/model"
)

var tagColorPattern = regexp.MustCompile(`^#[0-9A-Fa-f]{6}$`)

func (s *Store) Tags(ctx context.Context) ([]*model.Tag, error) {
	rows, err := s.q.ListTags(ctx, dbgen.ListTagsParams{})
	return dbutil.MapRows(rows, err, tagFromRow)
}

func (s *Store) TagByID(ctx context.Context, id int64) (*model.Tag, error) {
	rows, err := s.q.ListTags(ctx, dbgen.ListTagsParams{ID: &id})
	return dbutil.MapFirstRow(rows, err, tagFromRow)
}

func (s *Store) CreateTag(ctx context.Context, input model.CreateTagInput) (*model.Tag, error) {
	if err := validateTagColor(input.Color); err != nil {
		return nil, err
	}
	row, err := s.q.CreateTag(ctx, dbgen.CreateTagParams{Name: input.Name, Color: input.Color})
	if err != nil {
		return nil, err
	}
	return tagFromRow(row), nil
}

func (s *Store) UpdateTag(ctx context.Context, input model.UpdateTagInput) (*model.Tag, error) {
	if err := validateTagColor(input.Color); err != nil {
		return nil, err
	}
	id := input.ID.Int64()
	row, err := s.q.UpdateTag(ctx, dbgen.UpdateTagParams{ID: id, Name: input.Name, Color: input.Color})
	return dbutil.MapRow(row, err, tagFromRow)
}

func (s *Store) DeleteTag(ctx context.Context, id int64) (bool, error) {
	rows, err := s.q.DeleteTag(ctx, dbgen.DeleteTagParams{ID: id})
	return dbutil.RowsAffected(rows, err)
}

func (s *Store) TagsByTransactionIDs(ctx context.Context, transactionIDs []int64) (map[int64][]*model.Tag, error) {
	rows, err := s.q.TagsByTransactionIDs(ctx, dbgen.TagsByTransactionIDsParams{TransactionIds: transactionIDs})
	toTransactionTag := func(row dbgen.TagsByTransactionIDsRow) (int64, *model.Tag) {
		return row.TransactionID, tagFromRow(row.Tag)
	}
	return dbutil.GroupRows(rows, err, toTransactionTag)
}

func (s *Store) TagsByRuleIDs(ctx context.Context, ruleIDs []int64) (map[int64][]*model.Tag, error) {
	rows, err := s.q.TagsByRuleIDs(ctx, dbgen.TagsByRuleIDsParams{RuleIds: ruleIDs})
	toRuleTag := func(row dbgen.TagsByRuleIDsRow) (int64, *model.Tag) {
		return row.RuleID, tagFromRow(row.Tag)
	}
	return dbutil.GroupRows(rows, err, toRuleTag)
}

func tagFromRow(row dbgen.Tag) *model.Tag {
	return &model.Tag{
		ID:               model.New(model.GlobalIDTag, row.ID),
		Name:             row.Name,
		Color:            row.Color,
		TransactionCount: int32(row.TransactionCount),
	}
}

func validateTagColor(color string) error {
	if !tagColorPattern.MatchString(color) {
		return apierror.Publicf("tag color must be a #RRGGBB hex value")
	}
	return nil
}

File: server-rs/src/transactions/store/tags.rs
use std::collections::HashMap;

use anyhow::Result;
use sqlx::{Executor, Sqlite};

use crate::{apierror::ApiError, database::queries, transactions::Tag};

use super::mapping::{tag_from_fields, tag_from_row};

pub async fn tags(executor: impl Executor<'_, Database = Sqlite>) -> Result<Vec<Tag>> {
    queries::list_tags(executor, queries::ListTagsParams::default())
        .await
        .map(|rows| rows.into_iter().map(tag_from_row).collect())
        .map_err(Into::into)
}

pub async fn tag_by_id(executor: impl Executor<'_, Database = Sqlite>, id: i64) -> Result<Option<Tag>> {
    queries::list_tags(executor, queries::ListTagsParams { id: Some(id) })
        .await
        .map(|rows| rows.into_iter().next().map(tag_from_row))
        .map_err(Into::into)
}

pub async fn create_tag(executor: impl Executor<'_, Database = Sqlite>, name: &str, color: &str) -> Result<Tag> {
    validate_color(color)?;
    queries::create_tag(executor, queries::CreateTagParams { name, color })
        .await
        .map(|row| tag_from_fields(row.id, row.name, row.color, row.transaction_count))
        .map_err(Into::into)
}

pub async fn update_tag(
    executor: impl Executor<'_, Database = Sqlite>,
    id: i64,
    name: &str,
    color: &str,
) -> Result<Option<Tag>> {
    validate_color(color)?;
    queries::update_tag_opt(executor, queries::UpdateTagParams { name, color, id })
        .await
        .map(|row| row.map(|row| tag_from_fields(row.id, row.name, row.color, row.transaction_count)))
        .map_err(Into::into)
}

pub async fn delete_tag(executor: impl Executor<'_, Database = Sqlite>, id: i64) -> Result<bool> {
    queries::delete_tag(executor, queries::DeleteTagParams { id })
        .await
        .map(|rows| rows > 0)
        .map_err(Into::into)
}

pub async fn tags_by_transaction_ids(
    executor: impl Executor<'_, Database = Sqlite>,
    transaction_ids: &[i64],
) -> Result<HashMap<i64, Vec<Tag>>> {
    if transaction_ids.is_empty() {
        return Ok(HashMap::new());
    }
    queries::tags_by_transaction_ids(executor, queries::TagsByTransactionIDsParams { transaction_ids })
        .await
        .map(|rows| {
            rows.into_iter().fold(HashMap::<i64, Vec<Tag>>::new(), |mut tags, row| {
                tags.entry(row.transaction_id).or_default().push(tag_from_fields(
                    row.tags.id,
                    row.tags.name,
                    row.tags.color,
                    row.tags.transaction_count,
                ));
                tags
            })
        })
        .map_err(Into::into)
}

pub async fn tags_by_rule_ids(
    executor: impl Executor<'_, Database = Sqlite>,
    rule_ids: &[i64],
) -> Result<HashMap<i64, Vec<Tag>>> {
    if rule_ids.is_empty() {
        return Ok(HashMap::new());
    }
    queries::tags_by_rule_ids(executor, queries::TagsByRuleIDsParams { rule_ids })
        .await
        .map(|rows| {
            rows.into_iter().fold(HashMap::<i64, Vec<Tag>>::new(), |mut tags, row| {
                tags.entry(row.rule_id).or_default().push(tag_from_fields(
                    row.tags.id,
                    row.tags.name,
                    row.tags.color,
                    row.tags.transaction_count,
                ));
                tags
            })
        })
        .map_err(Into::into)
}

fn validate_color(color: &str) -> Result<()> {
    anyhow::ensure!(
        color.len() == 7 && color.starts_with('#') && color[1..].bytes().all(|byte| byte.is_ascii_hexdigit()),
        ApiError::bad_input("tag color must be a #RRGGBB hex value")
    );
    Ok(())
}

#[cfg(test)]
mod tests {
    use anyhow::Result;

    use super::*;

    #[tokio::test]
    async fn manages_tags_and_validates_colors() -> Result<()> {
        let pool = crate::database::dbtest::open().await?;
        let tag = create_tag(&pool, "Household", "#AABBCC").await?;
        assert_eq!(tag_by_id(&pool, tag.id).await?, Some(tag.clone()));
        assert!(tags(&pool).await?.iter().any(|item| item.id == tag.id));
        let updated = update_tag(&pool, tag.id, "Home", "#112233").await?.expect("tag exists");
        assert_eq!((updated.name.as_str(), updated.color.as_str()), ("Home", "#112233"));
        assert!(delete_tag(&pool, tag.id).await?);
        assert!(!delete_tag(&pool, tag.id).await?);
        assert_eq!(update_tag(&pool, tag.id, "Missing", "#112233").await?, None);
        assert_eq!(tags_by_transaction_ids(&pool, &[]).await?, HashMap::new());
        assert_eq!(tags_by_rule_ids(&pool, &[]).await?, HashMap::new());
        for color in ["#123", "112233", "#GGGGGG"] {
            assert_eq!(
                create_tag(&pool, "Invalid", color).await.unwrap_err().to_string(),
                "tag color must be a #RRGGBB hex value"
            );
        }
        Ok(())
    }
}
